src/Confluent.Kafka/Producer.cs

```diff
@@ -98,20 +522,135 @@ public class Producer<TKey, TValue> : ProducerBase, IProducer<TKey, TValue>
+            if (enableDeliveryReportsObj != null)
+            {
+                this.enableDeliveryReports = bool.Parse(enableDeliveryReportsObj.ToString());
+            }
+
+            var deliveryReportEnabledFieldsObj = config.FirstOrDefault(prop => prop.Key == ConfigPropertyNames.Producer.DeliveryReportFields).Value;
+            if (deliveryReportEnabledFieldsObj != null)
+            {
+                var fields = deliveryReportEnabledFieldsObj.ToString().Replace(" ", "");
+                if (fields != "all")
+                {
+                    this.enableDeliveryReportKey = false;
+                    this.enableDeliveryReportValue = false;
+                    this.enableDeliveryReportHeaders = false;
+                    this.enableDeliveryReportTimestamp = false;
+                    this.enableDeliveryReportPersistedStatus = false;
+                    if (fields != "none")
+                    {
+                        var parts = fields.Split(',');
+                        foreach (var part in parts)
+                        {
+                            switch (part)
+                            {
+                                case "key": this.enableDeliveryReportKey = true; break;
+                                case "value": this.enableDeliveryReportValue = true; break;
+                                case "timestamp": this.enableDeliveryReportTimestamp = true; break;
+                                case "headers": this.enableDeliveryReportHeaders = true; break;
+                                case "status": this.enableDeliveryReportPersistedStatus = true; break;
+                                default: throw new ArgumentException(
+                                    $"Unknown delivery report field name '{part}' in config value '{ConfigPropertyNames.Producer.DeliveryReportFields}'.");
+                            }
+                        }
+                    }
+                }
+            }
+
+            var configHandle = SafeConfigHandle.Create();
+
+            modifiedConfig.ToList().ForEach((kvp) => {
+                if (kvp.Value == null) throw new ArgumentException($"'{kvp.Key}' configuration parameter must not be null.");
+                configHandle.Set(kvp.Key, kvp.Value.ToString());
+            });
+
+
+            IntPtr configPtr = configHandle.DangerousGetHandle();
+
+            if (enableDeliveryReports)
+            {
+                Librdkafka.conf_set_dr_msg_cb(configPtr, DeliveryReportCallback);
+            }
+
+            // Explicitly keep references to delegates so they are not reclaimed by the GC.
+            errorCallbackDelegate = ErrorCallback;
+            logCallbackDelegate = LogCallback;
+            statisticsCallbackDelegate = StatisticsCallback;
+
+            // TODO: provide some mechanism whereby calls to the error and log callbacks are cached until
```

Comment:
This comment is no longer valid thanks to the builder pattern, right?

Reply:
yep, well spotted.